torch/_export/serde/serialize.py

```diff
@@ -633,15 +632,135 @@ def serialize(self, exported_program: ep.ExportedProgram) -> Tuple[ExportedProgr
         )
 
 
+class GraphModuleOpUpgrader:
+    """Given model op set version number as well as compiler op set version number, do the following:
+    1. retrieve upgraders from somewhere (TorchScript API or new API).
+    2. parse it and reorder for upgrading purpose.
+    3. register old versions of operators as custom ops.
+    4. prepare upgrader passes.
+
+    In `upgrade()` API run these upgrader passes.
+
+    An example of op_upgraders:
+    {
+        "aten::div__Scalar_0_3": (                              # versioned op name
+            "div._Scalar(self: Tensor, other: Scalar)",         # old schema
+            "
+def div__Scalar_0_3(self: Tensor, other: number) -> Tensor:     # upgrader in literal string
+  if (self.is_floating_point() or isinstance(other, float)):
+    return self.true_divide_(other)
+  return self.divide_(other, rounding_mode='trunc')
+",
+        ),
+    },
+    """
+
+    class UpgraderPass:
+        def __init__(self, old_target: Target, new_target: Target):
+            self.old_target = old_target
+            self.new_target = new_target
+
+        def __call__(self, fx_module: GraphModule) -> Optional[PassResult]:
+            replace_target_nodes_with(fx_module, "call_function", self.old_target, "call_function", self.new_target)
+            # retrace the graph module
+            return PassResult(fx_module, True)
+
+    def __init__(
+            self,
+            compiler_opset_version: Optional[Dict[str, int]] = None,
+            model_opset_version: Optional[Dict[str, int]] = None,
+            op_upgraders: Optional[Dict[str, Tuple[str, ...]]] = None,
+            # can add a new TS API: torch._C._get_upgraders_entry_map()
+    ):
+        self.compiler_opset_version = compiler_opset_version
+        self.model_opset_version = model_opset_version
+        # key is version number, value is a list of upgraders, keyed on op name
+        self.upgraders: List[Tuple[str]] = self._parse_upgraders(op_upgraders)
+
+        self.upgrader_passes: List[GraphModuleOpUpgrader.UpgraderPass] = self._populate_passes()
+
+    def _parse_upgraders(self, op_upgraders: Dict[str, Tuple[str]] = None) -> List[Tuple[str]]:
+        """reorder op_upgraders by version number."""
+        if not op_upgraders:
+            return []
+        model_ver = self.model_opset_version["aten"]
```

Comment:
what's special about `aten` here? If it's special, shouldn't we insist (i.e., early-validate) that opset versions always have `aten`?

Reply:
ATen op set versioning is something we had for TorchScript and I'm making sure it works in Export. We will add custom opset versioning in next PRs. But good point on validating it.